src/deriveNativeScriptHash.c

```diff
@@ -0,0 +1,464 @@
+#include "deriveNativeScriptHash.h"
+#include "state.h"
+#include "textUtils.h"
+#include "uiScreens.h"
+
+static ins_derive_native_script_hash_context_t* ctx = &(instructionState.deriveNativeScriptHashContext);
+
+// Helper functions
+
+#define TRACE_WITH_CTX(message, ...) TRACE(message "level = %u, remaining scripts = %u", ##__VA_ARGS__, ctx->level, ctx->complexScripts[ctx->level].remainingScripts)
+#define SAFE_SPRINTF(str, format, ...) {\
+		explicit_bzero(str, SIZEOF(str));\
+		snprintf(str, SIZEOF(str), format, __VA_ARGS__);\
+		ASSERT(strlen(str) + 1 < SIZEOF(str));\
+	}
+
+static inline bool areMoreScriptsExpected()
+{
+	// if the number of remaining scripts is not bigger than 0, then this request
+	// is invalid in the current context, as Ledger was not expecting another
+	// script to be parsed
+	return ctx->complexScripts[ctx->level].remainingScripts > 0;
+}
+
+static inline bool isComplexScriptFinished()
+{
+	return ctx->level > 0 && ctx->complexScripts[ctx->level].remainingScripts == 0;
+}
+
+static inline void complexScriptFinished()
+{
+	while (isComplexScriptFinished()) {
+		ASSERT(ctx->level > 0);
+		ctx->level--;
+
+		ASSERT(ctx->complexScripts[ctx->level].remainingScripts > 0);
+		ctx->complexScripts[ctx->level].remainingScripts--;
+
+		TRACE_WITH_CTX("complex script finished, ");
+	}
+}
+
+static inline void simpleScriptFinished()
+{
+	ASSERT(ctx->complexScripts[ctx->level].remainingScripts > 0);
+	ctx->complexScripts[ctx->level].remainingScripts--;
+
+	TRACE_WITH_CTX("simple script finished, ");
+
+	if (isComplexScriptFinished()) {
+		complexScriptFinished();
+	}
+}
+
+// UI
+
+const char ui_native_script_header[7][27] = {"Script - key path", "Script - key", "Script - ALL", "Script - ANY", "Script - N of K", "Script - invalid before", "Script - invalid hereafter"};
```

Comment:
This array should hold char*s, not whole char arrays, we can spare some memory with that if the average size of the strings is less than 23 (since the pointers themselves take up 4 bytes).

Reply:
I tried to change it, and it compiles, but Ledger crashes when working with these values. I am not sure why this happens, but now I remember that using fixed width here solved the issue for me. Maybe you will be wiser as to what is happening.

Reply:
Static strings are stored differently than code, and that leads to subtle problems with pointers. 
Take a look at
https://developers.ledger.com/docs/nano-app/position-independent-code/
There are macros to deal with it (namely, `PIC` and our own `PTR_PIC`, see `run_base58_test` for an example).

Reply:
https://github.com/vacuumlabs/ledger-app-cardano-shelley/issues/96